src/V3Ast.cpp

```diff
@@ -1114,6 +1114,31 @@ bool AstNode::sameTreeIter(const AstNode* node1p, const AstNode* node2p, bool ig
 //======================================================================
 // Debugging
 
+void AstNode::checkBitWidth(const AstNode* nodep) VL_MT_STABLE {
+    auto dumpSmall = [](const AstNodeExpr* exprp) VL_MT_STABLE {
+        std::ostringstream ss;
+        exprp->dtypep()->dumpSmall(ss);
+        return ss.str();
+    };
+    if (const AstNodeBiop* const biopp = VN_CAST(nodep, NodeBiop)) {
+        // Simple bitwise ops
+        if (VN_IS(biopp, And) || VN_IS(biopp, Or) || VN_IS(biopp, Xor)) {
```

Comment:
Perhaps instead setup the eventual framework for how this should be checked in the long run?
e.g. maybe have a virtual AstNodeBiop::widthBroken function that is called from here, then e.g.

bool AstAnd::widthBroken() override { return lhsp->widthMin() != rhsp->widthMin(); }

and call it from here?

Reply:
I think returning bool is not convenient, caller may want to know more in detail.
e.g. AstAnd/Or/Xor need to meet `lhsp->widthMin() == rhsp->widthMin() == this->widthMin()`.
bool is not enough to distinguish which widthMin is not correct.

Having UASSERT in such function gives more helpful error message.
```c++
virtual void AstAnd::checkWidth() const override { 
  UASSERT_OBJ(lhsp->widthMin() == rhsp()->widthMin(), this, ...);
  UASSERT_OBJ(lhsp->widthMin() == this->widthMin(), this, ...);
}
```
The code may be somewhat redundant, but informative.

Reply:
Ok, let's return a char* as to why, like the other broken() function, using the BROKEN_RTN function.

Reply:
OK. I guess it will be like below.
```c++
class AstNodeExpr  : ...{
    ...
  virtual const char * widthMismatch() const { return nullptr; } // declare virtual func in AstNodeExpr with default implementation
};

const char* AstAnd::widthMismatch() const {
  BROKEN_RTN(lhsp()->widthMin() != rhsp()->widthMin());
  BROKEN_RTN(this->widthMin() != lhsp()->widthMin());
  return nullptr;
}
```